proposals/016-dynamic-properties.md

```diff
@@ -0,0 +1,105 @@
+
+This functionality will give support to UI for implementing a pick list, auto-complete or property filter functionality. The initial implementation will focus on Salesforce use case, but it should provide enough room for other use cases.
+
+## User Story
+
+As a citizen user, I would like to pick property values instead of or in addition to manually inputting them in order to make it easier to select appropriate values or make fewer mistakes when selecting particular value in particular context.
+For example when using Salesforce's connector action _create or update_ I need to select a _Salesforce object type_ I wish to create or update and a _unique field_ of that type used to check for duplicates.
+
+## Domain
+
+The domain builds upon the _key-value_ model of properties that can be specified used to parametereize Actions. For a given property _key_ zero or more _values_ are provided based on the currently known context defined by connector, connection and already defined action property values.
+
+## API
+
+To interact with the filter API:
+
+| HTTP Verb | Path | Description |
+| --------- | ---- | ----------- |
+| PATCH     | /api/{version}/connectors/{connectorId}/actions/{actionId}/parameter/options | Lists available property values for given chosen property values |
+
+Swagger snippet:
+
+```yaml
+/connectors/{connectorId}/actions/{actionId}/parameter/options:
+  patch:
+    tags:
+    - "connectors"
+    summary: "Lists available property values"
+    description: "Returns a key - value-list of properties that can be specified\
+      \ for already choosen properties."
+    operationId: "getParameterOptions"
+    consumes:
+    - "application/json"
+    produces:
+    - "application/json"
+    parameters:
+    - name: "connectorId"
+      in: "path"
+      required: true
+      type: "string"
+      x-example: "salesforce"
+    - name: "actionId"
+      in: "path"
+      required: true
+      type: "string"
+      x-example: "io.syndesis:salesforce-create-or-update:latest"
+    - in: "body"
+      name: "choosenProperties"
+      required: false
+      schema:
+        type: "object"
+        additionalProperties:
+          type: "string"
+      x-examples:
+        application/json: "{\"sObjectName\": \"Account\"}"
+    responses:
+      200:
+        description: "Set of property key-value pairs that can be specified. The\
+          \ list of values is filtered by the given `choosenProperties` to contain\
+          \ only the values that make sense for those."
+        schema:
+          type: "object"
+          additionalProperties:
+            $ref: "#/definitions/ActionParameterValue"
+```
+
+### Example
+
+If the user is working with Salesforce Action _create or update_ Salesforce object and has already picked the _Contact_ Salesforce object to create or update and whishes to determine possible values for the Salesforce object unique ID field, the following request is issued:
```

Comment:
Another question (not sure whether related to the story, please stop me, if not ;-): What about other input type specific for the action verb/object, e.g. plain text fields ? Should these be a also returned by such a service, so that the UI could use a kind of generic form builder to create all the data required ?